Python/bytecodes.c

```diff
@@ -4394,16 +4399,27 @@ dummy_func(
             _GUARD_CALLABLE_ISINSTANCE +
             _CALL_ISINSTANCE;
 
+        macro(CALL_LIST_APPEND) =
+            unused/1 +
+            unused/2 +
+            _GUARD_CALLABLE_LIST_APPEND +
+            _GUARD_NOS_NOT_NULL +
+            _GUARD_NOS_LIST +
+            _CALL_LIST_APPEND;
+
+        op(_GUARD_CALLABLE_LIST_APPEND, (callable, unused, unused -- callable, unused, unused)){
+            PyObject *callable_o = PyStackRef_AsPyObjectBorrow(callable);
+            PyInterpreterState *interp = tstate->interp;
+            DEOPT_IF(callable_o != interp->callable_cache.list_append);
+        }
+
         // This is secretly a super-instruction
-        inst(CALL_LIST_APPEND, (unused/1, unused/2, callable, self, arg -- )) {
+        op(_CALL_LIST_APPEND, (callable, self, arg -- )) {
             assert(oparg == 1);
-            PyObject *callable_o = PyStackRef_AsPyObjectBorrow(callable);
             PyObject *self_o = PyStackRef_AsPyObjectBorrow(self);
 
-            PyInterpreterState *interp = tstate->interp;
-            DEOPT_IF(callable_o != interp->callable_cache.list_append);
             DEOPT_IF(self_o == NULL);
-            DEOPT_IF(!PyList_Check(self_o));
```

Comment:
Wait a minute, the old check looks bugged? It should be an exact check not a subtype check. If I understand correctly, we should fix this on 3.13/3.14 too.

Can you check with someone else to confirm this please?

Reply:
No, it's okay. The list append function works with subclasses. The old code was more permissive, allowing `my_list_subclass.append(foo)`. We don't allow that anymore (in order to narrow an exact type) but the old code wasn't *wrong*.

Reply:
Oh wait yeah I didn't see the `DEOPT_IF(callable_o != interp->callable_cache.list_append);` line above. It's safe then. Without that line it would be wrong I think (due to the possibility of the append being overriden).